Make torch.linalg.svdvals differentiable

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57189 Make torch.linalg.eigvalsh differentiable
* **#57188 Make torch.linalg.svdvals differentiable**

`torch.linalg.svdvals` now supports autograd. This is achieved by
computing the singular vectors internally if input requires grad,
otherwise the vectors are not computed and the operation is faster.

Differential Revision: [D28199709](https://our.internmc.facebook.com/intern/diff/D28199709)

IvanYashchuk committed:
Make torch.linalg.svdvals differentiable

`torch.linalg.svdvals` now supports autograd. This is achieved by
computing the singular vectors internally if input requires grad,
otherwise the vectors are not computed and the operation is faster.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Make torch.linalg.svdvals differentiable"


`torch.linalg.svdvals` now supports autograd. This is achieved by
computing the singular vectors internally if input requires grad,
otherwise the vectors are not computed and the operation is faster.

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2763,10 +2764,20 @@ Tensor linalg_svdvals(const Tensor& input) {
+    std::tie(std::ignore, singular_values, std::ignore) =
+        // NOLINTNEXTLINE(bugprone-argument-comment)
+        at::_svd_helper(input, /*!full_matrices=*/false, /*compute_uv=*/false);
+  }
```
Perhaps is simpler to avoid repeating code:
```c++
const auto needs_grad = torch::autograd::compute_requires_grad(input);
return std::get<1>(at::_svd_helper(input, /*!full_matrices=*/needs_grad, /*compute_uv=*/needs_grad));
``` 
Also, can't we just pass `!full_matrices = True` in both cases?

IvanYashchuk replied:
`!full_matrices` doesn't do anything if `compute_uv=False`, so it doesn't matter, it can be true in both code paths.

If the singular vectors are computed in full matrix form, then there is some additional work in the derivative function
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L2066-L2078

Lezcano replied:
That's what I thought. Then, I update my proposal to:
```c++
const auto compute_uv = torch::autograd::compute_requires_grad(input);
return std::get<1>(at::_svd_helper(input, /*!full_matrices=*/True, /*compute_uv=*/compute_uv));
```

IvanYashchuk replied:
I find std::tie with std::ignore more clear than std::get<1>. I can of course change that, it doesn't matter now.

Lezcano replied:
Whichever works best. I'd still remove the if/else logic though

lezcano reviewed (commented):
Also just a very minor point.

I think this would also be a good time to add a comment like that in `eig` and `eigh` in the `seealso` section. (cf. https://github.com/pytorch/pytorch/pull/57189#discussion_r622874728)

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1298,9 +1298,6 @@
 
-.. note:: This function is not differentiable. If you need differentiability use
-          :func:`torch.linalg.svd` instead, which also computes the singular vectors.
-
```
Same as https://github.com/pytorch/pytorch/pull/57189#discussion_r622914613

mruberry reviewed (approved):
Cool! I also like @Lezcano's suggestion to make this consistent with the next PR in the stack

@albanD would you take a look and, once you're happy and @Lezcano's suggestions are addressed (or resolved) land this stack?

albanD commented:
Yes, waiting on the doc update and then I can take care of landing this stack.

mruberry commented:
Note: this may be blocking removal of compute_uv, since one of the internal users of compute_uv=False wants differentiability

IvanYashchuk commented:
`compute_uv=False` is not differentiable currently and it will not be. Because in order to differentiate the singular values, singular vectors must be computed.

IvanYashchuk commented:
Running `pytest test/test_ops.py -k 'svdvals' --durations=10` gives
```
========================================================== slowest 10 durations ==========================================================
4.91s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_svdvals_cuda_complex64
4.82s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_svdvals_cuda_float32
4.49s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_svdvals_cuda_complex128
4.00s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_svdvals_cpu_complex64
3.59s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_svdvals_cpu_float32
2.58s call     test/test_ops.py::TestOpInfoCUDA::test_duplicate_method_tests_linalg_svdvals_cuda_float32
1.84s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_svdvals_cuda_float64
1.61s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_svdvals_cpu_complex128
1.55s call     test/test_ops.py::TestOpInfoCUDA::test_supported_backward_linalg_svdvals_cuda_complex128
0.92s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_svdvals_cuda_complex128

```

IvanYashchuk committed:
Update on "Make torch.linalg.svdvals differentiable"


`torch.linalg.svdvals` now supports autograd. This is achieved by
computing the singular vectors internally if input requires grad,
otherwise the vectors are not computed and the operation is faster.

[ghstack-poisoned]

IvanYashchuk commented:
@albanD, I updated the documentation in both PRs and I modified the implementation in this PR according to Mario's suggestion. This stack is ready to be merged.

albanD commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -10,6 +10,7 @@
 #include <ATen/native/cpu/zmath.h>
 #include <ATen/Parallel.h>
 
+#include <torch/csrc/autograd/functions/utils.h>
```
I'm afraid you can't do that... libtorch depends on aten, but not the other way around...

albanD replied:
I guess you can work around this here by checking `GradMode::is_enabled() && input.requires_grad()` below instead of using this function...

IvanYashchuk replied:
Alright, I will change it then.

IvanYashchuk replied:
@albanD, done.

IvanYashchuk committed:
Update on "Make torch.linalg.svdvals differentiable"


`torch.linalg.svdvals` now supports autograd. This is achieved by
computing the singular vectors internally if input requires grad,
otherwise the vectors are not computed and the operation is faster.

Differential Revision: [D28199709](https://our.internmc.facebook.com/intern/diff/D28199709)

[ghstack-poisoned]
